Make start date and day of month aware of timezones (fixes #12555)

bakerty opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #12555

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Bug was caused by updateTask api method setting `daysOfMonth = startDate.date()` and because `startDate` is represented in UTC, `daysOfMonth` would be wrong if user was in a UTC+xxx timezone.

Fixed this by factoring in `user.preferences.timezoneOffset`, which sets `daysOfMonth` to reflect the date in the user's timezone instead of UTC.

Also slightly tweaked the test for this api call so that the updated task start date was different from the original.


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 76ad7ada-553e-46bc-a6dc-fcead49f9228

paglias commented:
Hi @bakerty , thanks for the PR.

I noticed this is also changing the value for `startDate`, I guess this is being done because it's not set correctly at task creation, maybe it should be fixed there instead of when updating the task?

bakerty commented:
Thanks for the response @paglias ,

`startDate` gets set properly on task creation. The reason for the change is that when the task gets updated with a new `startDate`, only the date changes and not the time, which could potentially cause a problem if the user changes timezones. This also only affects 'day of month' dailys

If this is actually the intended behaviour then I apologize and I'd be happy to remove that section of code and re-submit.

paglias commented:
Hmm, I'm going to ping @Alys to have an extra pair of eyes on this

Alys commented:
I believe changing the Start Date is good!

We used to tell people to change the Start Date to fix Dailies that had become rescheduled onto a different day by a time zone change, but I think that stopped working (fairly recently?). From what I can see, the change in this PR is good.

I wouldn't mind testing it on our delta test site for a few days if that's okay @paglias ?

paglias commented:
My point was that if there's a problem with the start date not being set correctly, it should be fixed at task creation, not later. Because right now it's not being set this way on creation. @Alys I'm putting this on delta

paglias commented:
@Alys have you had a chance to look at this?

Alys commented:
No. The past couple of weeks have had more RaB posts than usual and I've been helping with the excess emails. Testing this is still on my task list.

Alys commented:
@bakerty Sorry about the delay with my testing! I've tested this on our delta test site and it's great. It fixes https://github.com/HabitRPG/habitica/issues/12555 and changing the value for `startDate` works as you've described, and allows this workaround to be used again:

> We used to tell people to change the Start Date to fix Dailies that had become rescheduled onto a different day by a time zone change, but I think that stopped working

@paglias I'd be happy for this to go live whenever you think best.

paglias commented:
Thanks @Alys , before we merge this @bakerty I'd like the fix to be applied on task creation as well since it seems obvious that it's not accounting for time zone there, is this something you can add as well? Thanks

Alys commented:
> I'd like the fix to be applied on task creation as well since it seems obvious that it's not accounting for time zone there

Hmmm, yes. When you create a Daily through the website, your time zone IS accounted for and the Start Date has the correct time in it. When you create a Daily through the API, your time zone is NOT used when setting the Start Date. I didn't realise until now that the API wasn't doing that. This may account for some of the scheduling bugs we see in the mobile apps.

bakerty commented:
Alright I'll put the change in at task creation as well. Thank you both for all the feedback!

Should I extend the scope of the `startDate` modification to include all dailys, instead of only monthly ones, as well?

paglias commented:
I'd say all dailies would be better

bakerty committed:
WIP: #12555

-dayOfMonth takes timezone into account on task update for monthly dailys
-startDate set to start of day in user's tz on task update for monthly
dailys
-tweaked a test so that it's actually testing something

bakerty committed:
WIP: 12555

-task.startDate gets set to start of day in user's tz on task creation
and task update
-set preferences.timezoneOffset to nonzero value in certian test user objects
-removed date literals in test:api-v3:integration:tasks:PUT

bakerty commented:
Hey @Alys or @paglias, I'm looking for some advice here.

I added some code to `createTasks` in `website/server/libs/taskManager.js` that sets `startDate` to midnight in the user's timezone on task creation, and tweaked the relevant tests to test the new behaviour. All of it seems to be working as expected.

However, I ran into a rather major problem. If the user selects a start date far enough in the future, then it's possible that the user's timezone will change before that date thanks to daylight savings. This would mean that `startDate` might be inaccurate by 1 hour in either direction. The same problem occurs for monthly dailys, where some dates in `[nextDue]` are inaccurate because of daylight savings. This causes some test to not pass if they are run in a location where daylight savings happens.

The main issue is that it's impossible to figure out if a submitted start date is meant to be in daylight savings time or standard time server-side. I'm not really sure what the best way to proceed with this problem is. Any help would be greatly appreciated.

paglias commented:
Hmm let's see what @Alys thinks but i'd say that if it fixes the issue in most cases we should be fine, it's still an improvement on the current situation

paglias commented:
@Alys any update on this one?

Alys commented:
Sorry! I missed this and PR 12860 in the pre-Christmas chaos. I'll look through both this weekend.

Alys commented:
I think this is good as is!

> If the user selects a start date far enough in the future, then it's possible that the user's timezone will change before that date thanks to daylight savings. This would mean that startDate might be inaccurate by 1 hour in either direction.

That's definitely a problem but it always has been, it isn't made worse by this PR, and it's not something easily fixed. I guess we'd have to add a separate time zone field to each task to record the zone when it was created / modified, and then consider that when scheduling the task, but that's far out of scope for this issue.

@bakerty Thanks very much for this fix and I'm so sorry it took me this long to reply!

paglias commented:
@bakerty  thanks for this PR! I'm merging the PR and awarding your contributor tier.

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
